### DISCUSSION TITLE
An error occurred (InvalidParameterException) when calling the UpdateManagedLoginBranding operation: Media type not supported for value: text/plain

### DISCUSSION THREAD

**Discussion by ghost** [aws/aws-cli#9102] [Q&A]
When trying to add assets to **aws cognito-idp update-managed-login-branding** or **aws cognito-idp create-managed-login-branding**, I can't seem to find the proper way to upload the asset.

https://docs.aws.amazon.com/cli/latest/reference/cognito-idp/update-managed-login-branding.html says the file type should be "blob" and I've tried taking the asset output (byte string) from a "describe-managed-login-branding" request and additionally tried adding a file path as shown here: https://docs.aws.amazon.com/cli/latest/userguide/cli-usage-parameters-types.html#parameter-type-blob and seem to always get the error: **An error occurred (InvalidParameterException) when calling the UpdateManagedLoginBranding operation: Media type not supported for value: text/plain**

If I don't include assets in the command, everything works fine.

I've tried:

aws cognito-idp update-managed-login-branding --user-pool-id "USERPOOL" --managed-login-branding-id "BRANDING" --settings "$(< ./settings.json)" --assets **"$(< ./assets.json)"**

aws cognito-idp update-managed-login-branding --user-pool-id "USERPOOL" --managed-login-branding-id "BRANDING" --settings "$(< ./settings.json)" --assets '[{"Category": "FORM_LOGO", "ColorMode": "DARK", "Extension": "PNG", **"Bytes": "file://logo.png"**}]'

aws cognito-idp update-managed-login-branding --user-pool-id "USERPOOL" --managed-login-branding-id "BRANDING" --settings "$(< ./settings.json)" --assets '[{"Category": "FORM_LOGO", "ColorMode": "DARK", "Extension": "PNG", **"Bytes": "fileb://logo.png"**}]'

aws cognito-idp update-managed-login-branding --user-pool-id "USERPOOL" --managed-login-branding-id "BRANDING" --settings "$(< ./settings.json)" --assets '[{"Category": "FORM_LOGO", "ColorMode": "DARK", "Extension": "PNG", **"Bytes": "./logo.png"**}]'

aws cognito-idp update-managed-login-branding --user-pool-id "USERPOOL" --managed-login-branding-id "BRANDING" --settings "$(< ./settings.json)" --assets '[{"Category": "FORM_LOGO", "ColorMode": "DARK", "Extension": "PNG", **"Bytes": "$(cat ./logo.png | base64)"**}]'

aws cognito-idp update-managed-login-branding --user-pool-id "USERPOOL" --managed-login-branding-id "BRANDING" --settings "$(< ./settings.json)" --assets '[{"Category": "FORM_LOGO", "ColorMode": "DARK", "Extension": "PNG", **"Bytes": null**}]' - which returned **Invalid type for parameter Assets[0].Bytes, value: None, type: <class 'NoneType'>, valid types: <class 'bytes'>, <class 'bytearray'>, file-like object**

**Comment by vgul** [comment]
I have the same issue

**Comment by uldyssian-sh** [comment]
That error happens because `Bytes` is a *blob* field, but when you put `file://...` or `fileb://...` inside JSON, AWS CLI treats it as a literal string and sends the text `fileb://logo.png` to the service.

Cognito then tries to interpret the “bytes” you sent and detects it as plain text (`text/plain`), which is why it rejects it with “Media type not supported”.

To make AWS CLI actually read the binary file and send it as a blob, you must pass the blob value via CLI argument parsing (where `fileb://` is supported), not as a JSON string.

Use shorthand for `--assets` (this is the key):

aws cognito-idp update-managed-login-branding \
  --user-pool-id "USERPOOL" \
  --managed-login-branding-id "BRANDING" \
  --settings file://settings.json \
  --assets "Category=FORM_LOGO,ColorMode=DARK,Extension=PNG,Bytes=fileb://logo.png"


If you want to keep everything in JSON, then `Bytes` must be an actual Base64 string in the JSON (not `fileb://...`). For example, generate Base64 and put it into the JSON yourself:

BASE64=$(base64 -i logo.png | tr -d '\n')

Then in your JSON request:

{
  "UserPoolId": "USERPOOL",
  "ManagedLoginBrandingId": "BRANDING",
  "Settings": { ... },
  "Assets": [
    {
      "Category": "FORM_LOGO",
      "ColorMode": "DARK",
      "Extension": "PNG",
      "Bytes": "BASE64_STRING_HERE"
    }
  ]
}

Then call:

aws cognito-idp update-managed-login-branding --cli-input-json file://request.json


In short: `fileb://...` works for blob parameters when passed as an argument value, but it will not be expanded when embedded inside JSON; embedding it makes Cognito receive plain text, leading to the `text/plain` media type error.
